### PR TITLE
CONSOLE-2501: Upgrade to Typescript v4

### PR DESCRIPTION
**Pull request by glekner**
- Updated Major typescript version to match the dynamic plugin sdk version `4.2.4`
- Updated `ts-loader` and `ForkTsCheckerWebpackPlugin` versions. cli logs about ts errors are now a lot clearer.
- Updated `i18n-parser` to support ts4
- Updated `@typescript-eslint/eslint-plugin` and `@typescript-eslint/parser` from v2 to v5
- Updated `eslint-plugin-react-hooks` to v4 - Added lots of missing `useMemo`s
- Added globals like `React` `JSX` 
- Fixed breaking changes and rules
- Changed `Chainable<Subject>` to `Chainable` (cypress). eslint would shout about unused vars, and since updating versions `Subject` is the default type `interface Cypress.Chainable<Subject = any>`

CLI Errors:
Before
Screen Shot 2021-11-10 at 19 42 18

After
Screen Shot 2021-11-10 at 19 35 50

### CODE REVIEW

**Comment by glekner** [comment]
please review @spadgett @vojtechszocs @rawagner @christianvogt

File: frontend/package.json [hunk 1]
```diff
@@ -264,14 +264,14 @@
...
-    "fork-ts-checker-webpack-plugin": "4.x",
+    "fork-ts-checker-webpack-plugin": "6.x",
     "geckodriver": "1.x",
     "glob": "7.x",
     "glslify-loader": "^2.0.0",
     "graphql-tag": "^2.10.3",
     "html-webpack-plugin": "3.x",
     "husky": "^4.2.5",
-    "i18next-parser": "^3.3.0",
+    "i18next-parser": "^3.x",
```
Comment (spadgett) [inline frontend/package.json:274]: `3.x` makes sense to use with `^`. We should use one or the other.

File: frontend/package.json [hunk 2]
```diff
@@ -305,9 +305,9 @@
     "style-loader": "^2.0.0",
     "thread-loader": "^3.0.4",
     "ts-jest": "21.x",
-    "ts-loader": "^6.2.2",
+    "ts-loader": "8.3.0",
```
Comment (spadgett) [inline frontend/package.json:308]: This should have a `~` or a `^`. I don't think we need to lock to a specific version.

File: frontend/packages/ceph-storage-plugin/src/components/create-storage-system/external-storage/ibm-flashsystem/index.tsx [hunk 3]
```diff
@@ -166,7 +166,6 @@ export const createFlashSystemPayload: CreatePayload<FlashSystemState> = (
   const storageSecretTemplate: SecretKind = {
     apiVersion: apiVersionForModel(SecretModel),
     stringData: {
-      // eslint-disable-next-line @typescript-eslint/camelcase
```
Comment (spadgett) [inline frontend/packages/ceph-storage-plugin/src/components/create-storage-system/external-storage/ibm-flashsystem/index.tsx:169]: Hm, I don't see why this comment could be removed. We want the camelcase rule active for most of the code. Did the defaults change?

File: frontend/packages/console-shared/src/components/popper/Popper.tsx [hunk 4]
```diff
@@ -25,10 +25,10 @@ class VirtualReference implements PopperJSReference {
       right: x + width,
       top: y,
       width,
-    };
+    } as DOMRect;
```
Comment (spadgett) [inline frontend/packages/console-shared/src/components/popper/Popper.tsx:28]: Why is `as` needed here?

File: frontend/packages/dev-console/src/components/catalog/providers/useBuilderImages.tsx [hunk 5]
```diff
@@ -29,13 +29,13 @@ const normalizeBuilderImages = (
...
     const iconClass = imgUrl ? null : icon;
-    const description = tag?.['annotations']?.['description'] ?? '';
+    const description = tag?.annotations?.description ?? '';
     const tags = getAnnotationTags(tag);
     const createLabel = t('devconsole~Create Application');
     const provider = annotations?.[ANNOTATIONS.providerDisplayName] ?? '';
     const href = `/catalog/source-to-image?imagestream=${name}&imagestream-ns=${namespace}&preselected-ns=${activeNamespace}`;
-    const builderImageTag = _.head(imageStream.spec?.tags);
-    const sampleRepo = builderImageTag?.['annotations']?.['sampleRepo'];
+    const builderImageTag = _.head<any>(imageStream.spec?.tags);
```
Comment (spadgett) [inline frontend/packages/dev-console/src/components/catalog/providers/useBuilderImages.tsx:37]: Tags should be a `string[]`.

File: frontend/packages/dev-console/src/components/import/image-search/ImageStreamDropdown.tsx [hunk 6]
```diff
@@ -13,6 +13,7 @@ const ImageStreamDropdown: React.FC<{ disabled?: boolean; formContextField?: str
   formContextField,
 }) => {
   const { t } = useTranslation();
+  // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (spadgett) [inline frontend/packages/dev-console/src/components/import/image-search/ImageStreamDropdown.tsx:16]: Why is this needed here?

File: frontend/packages/eslint-plugin-console/lib/config/rules/typescript.js [hunk 7]
```diff
@@ -12,17 +12,23 @@ module.exports = {
...
+        '{}': false,
+        Function: false,
+        object: false,
+      },
+    },
+  ],
 
   // Enforce camelCase naming convention
   camelcase: 'off',
-  '@typescript-eslint/camelcase': 'error',
```
Comment (spadgett) [inline frontend/packages/eslint-plugin-console/lib/config/rules/typescript.js:22]: Why disable this?

File: frontend/packages/eslint-plugin-console/lib/config/rules/typescript.js [hunk 8]
```diff
@@ -12,17 +12,23 @@ module.exports = {
...
-
-  // Require PascalCased class and interface names
-  '@typescript-eslint/class-name-casing': 'error',
```
Comment (spadgett) [inline frontend/packages/eslint-plugin-console/lib/config/rules/typescript.js:25]: Why disable this?

File: frontend/packages/eslint-plugin-console/lib/config/rules/typescript.js [hunk 9]
```diff
@@ -162,7 +168,7 @@ module.exports = {
 
   // Disallow the use of variables before they are defined
   'no-use-before-define': 'off',
-  '@typescript-eslint/no-use-before-define': 'error',
+  '@typescript-eslint/no-use-before-define': 'off',
```
Comment (spadgett) [inline frontend/packages/eslint-plugin-console/lib/config/rules/typescript.js:171]: Why change this?

File: frontend/packages/operator-lifecycle-manager/src/components/modals/uninstall-operator-modal.tsx [hunk 10]
```diff
@@ -222,30 +214,7 @@ export const UninstallOperatorModal: React.FC<UninstallOperatorModalProps> = ({
     </>
   );
 
-  // eslint-disable-next-line @typescript-eslint/no-unused-vars,@typescript-eslint/ban-ts-ignore
-  // @ts-ignore
   // eslint-disable-next-line @typescript-eslint/no-unused-vars
-  const operandsSection = operandsLoadedErrorMessage ? (
```
Comment (spadgett) [inline frontend/packages/operator-lifecycle-manager/src/components/modals/uninstall-operator-modal.tsx:228]: We're about to reenable this, so we shouldn't delete the code. Let's comment out instead.

File: frontend/packages/patternfly/.eslintrc [hunk 11]
```diff
@@ -45,13 +45,12 @@
     "no-else-return": ["error"],
     "no-global-strict": 0,
     "no-irregular-whitespace": ["error"],
-    "no-shadow": ["error"],
```
Comment (spadgett) [inline frontend/packages/patternfly/.eslintrc:48]: Why the change?

File: frontend/packages/patternfly/.eslintrc [hunk 12]
```diff
@@ -45,13 +45,12 @@
...
     "no-underscore-dangle": 0,
     "@typescript-eslint/no-unused-vars": [
       "error",
       { "varsIgnorePattern": "React", "args": "after-used" }
     ],
-    "@typescript-eslint/no-use-before-define": 2,
+    "@typescript-eslint/no-use-before-define": "off",
```
Comment (spadgett) [inline frontend/packages/patternfly/.eslintrc:53]: Why the change?

File: frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-metrics/charts/TimeSeriesChart.tsx [hunk 13]
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable dot-notation */
```
Comment (spadgett) [inline frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-metrics/charts/TimeSeriesChart.tsx:1]: Can we fix instead with `--fix`?

File: frontend/public/components/RBAC/edit-rule.jsx [hunk 14]
```diff
@@ -297,6 +297,7 @@ const EditRule = connect(stateToProps, { getResources: k8sActions.getResources }
     }
 
     render() {
+      // eslint-disable-next-line react-hooks/rules-of-hooks
```
Comment (spadgett) [inline frontend/public/components/RBAC/edit-rule.jsx:300]: We should open a bug to fix this.

File: frontend/public/components/dashboard/dashboards-page/cluster-dashboard/inventory-card.tsx [hunk 15]
```diff
@@ -79,6 +79,7 @@ const ClusterInventoryItem = withDashboardResources<ClusterInventoryItemProps>(
             });
       }, [mapperLoader]);
 
+      // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (spadgett) [inline frontend/public/components/dashboard/dashboards-page/cluster-dashboard/inventory-card.tsx:82]: Why was this needed?

File: frontend/public/components/dashboard/dashboards-page/cluster-dashboard/utilization-card.tsx [hunk 16]
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable react-hooks/exhaustive-deps */
```
Comment (spadgett) [inline frontend/public/components/dashboard/dashboards-page/cluster-dashboard/utilization-card.tsx:1]: Let's avoid adding this at the top of the file. It's better to apply only where it's needed.

File: frontend/public/components/dashboard/project-dashboard/utilization-card.tsx [hunk 17]
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable react-hooks/exhaustive-deps */
```
Comment (spadgett) [inline frontend/public/components/dashboard/project-dashboard/utilization-card.tsx:1]: Same comment here.

File: frontend/public/components/filter-toolbar.tsx [hunk 18]
```diff
@@ -101,6 +101,7 @@ export const FilterToolbar: React.FC<FilterToolbarProps> = ({
   const labelFilterQueryArgumentKey = uniqueFilterName
     ? `${uniqueFilterName}-${labelFilter}`
     : labelFilter;
+  // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (spadgett) [inline frontend/public/components/filter-toolbar.tsx:104]: Why was this needed?

File: frontend/public/components/monitoring/hooks/useIsVisible.ts [hunk 19]
```diff
@@ -11,6 +11,7 @@ export const useIsVisible = (ref) => {
     }
   };
 
+  // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (spadgett) [inline frontend/public/components/monitoring/hooks/useIsVisible.ts:14]: Why was this needed?

**Comment by spadgett** [comment]
We'll need to solve this error:

> error @typescript-eslint/eslint-plugin@5.3.1: The engine "node" is incompatible with this module. Expected version "^12.22.0 || ^14.17.0 || >=16.0.0". Got "14.16.0"

File: frontend/packages/console-shared/src/components/popper/Popper.tsx (hunk 4 above)
Comment (glekner) [reply frontend/packages/console-shared/src/components/popper/Popper.tsx:28]: resolves
`Type '{ bottom: number; height: number; left: number; right: number; top: number; width: number; }' is missing the following properties from type 'DOMRect': x, y, toJSONts(2739)`

File: frontend/public/components/dashboard/dashboards-page/cluster-dashboard/inventory-card.tsx (hunk 15 above)
Comment (glekner) [reply frontend/public/components/dashboard/dashboards-page/cluster-dashboard/inventory-card.tsx:82]: resolves
`The 'additionalResourcesData' object makes the dependencies of useCallback Hook (at line 105) change on every render. To fix this, wrap the initialization of 'additionalResourcesData' in its own useMemo()`

no need for useMemo here

File: frontend/packages/eslint-plugin-console/lib/config/rules/typescript.js (hunk 9 above)
Comment (glekner) [reply frontend/packages/eslint-plugin-console/lib/config/rules/typescript.js:171]: the linter is smarter now it seems, a lot of cases where functions where used before defined.

**Comment by glekner** [comment]
@spadgett 
- now upgrading @typescript-eslint/eslint-plugin and @typescript-eslint/parser to v4 instead of v5. v5 is mainly for ESLint 8 which we don't use. our node version is now supported
- camelcase is deprecated in favor of https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/naming-convention.md
- Note, all useCallback eslint disabling is because the plugin prefer inline functions instead of external ones even though they are completely fine.
- `cx` usages also needed eslint disabling for `naming-convention` because of the classnames

File: frontend/packages/console-shared/src/components/popper/Popper.tsx (hunk 4 above)
Comment (spadgett) [reply frontend/packages/console-shared/src/components/popper/Popper.tsx:28]: That seems like a legitimate type error we should address... Either `DOMRect` is the wrong type or we're not specifying all the properties it should have.

File: frontend/packages/eslint-plugin-console/lib/config/rules/typescript.js (hunk 9 above)
Comment (spadgett) [reply frontend/packages/eslint-plugin-console/lib/config/rules/typescript.js:183]: We can disable just functions if we need to:

https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-use-before-define.md#other-options

https://eslint.org/docs/rules/no-use-before-define#functions

File: frontend/public/components/dashboard/dashboards-page/cluster-dashboard/inventory-card.tsx (hunk 15 above)
Comment (spadgett) [reply frontend/public/components/dashboard/dashboards-page/cluster-dashboard/inventory-card.tsx:82]: That sounds like a legitimate problem that shouldn't be ignored?

**Comment by rebeccaalpert** [comment]
```
/go/src/github.com/openshift/console/frontend/node_modules/ts-node/src/index.ts:750
    return new TSError(diagnosticText, diagnosticCodes);
           ^
TSError: ⨯ Unable to compile TypeScript:
scripts/generate-pkg-assets.ts(20,10): error TS2550: Property 'entries' does not exist on type 'ObjectConstructor'. Do you need to change your target library? Try changing the `lib` compiler option to 'es2017' or later.
    at createTSError (/go/src/github.com/openshift/console/frontend/node_modules/ts-node/src/index.ts:750:12)
    at reportTSError (/go/src/github.com/openshift/console/frontend/node_modules/ts-node/src/index.ts:754:19)
    at getOutput (/go/src/github.com/openshift/console/frontend/node_modules/ts-node/src/index.ts:941:36)
    at Object.compile (/go/src/github.com/openshift/console/frontend/node_modules/ts-node/src/index.ts:1243:30)
    at Module.m._compile (/go/src/github.com/openshift/console/frontend/node_modules/ts-node/src/index.ts:1370:30)
    at Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
    at Object.require.extensions.<computed> [as .ts] (/go/src/github.com/openshift/console/frontend/node_modules/ts-node/src/index.ts:1374:12)
    at Module.load (internal/modules/cjs/loader.js:928:32)
    at Function.Module._load (internal/modules/cjs/loader.js:769:14)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:72:12) {
  diagnosticText: "scripts/generate-pkg-assets.ts(20,10): error TS2550: Property 'entries' does not exist on type 'ObjectConstructor'. Do you need to change your target library? Try changing the `lib` compiler option to 'es2017' or later.\n",
  diagnosticCodes: [ 2550 ]
```

**Comment by rebeccaalpert** [comment]
I spoke to Kim and she suggested creating a branch based on glekner:ts4/a new PR and then close this. I'll take a look at the GH comments before I do that to make sure they have been addressed first.

File: frontend/packages/console-shared/src/components/popper/Popper.tsx (hunk 4 above)
Comment (rebeccaalpert) [reply frontend/packages/console-shared/src/components/popper/Popper.tsx:28]: I'm going to back this change out in the new PR and see what's needed to resolve the type error.

File: frontend/packages/dev-console/src/components/import/image-search/ImageStreamDropdown.tsx (hunk 6 above)
Comment (rebeccaalpert) [reply frontend/packages/dev-console/src/components/import/image-search/ImageStreamDropdown.tsx:16]: I will remove this and see what happens.

**Comment by glekner** [comment]
@rebeccaalpert I fixed the TS compilation error.

The only remaining issue AFAIK is jest. I assume the current version of `ts-jest` doesn't support TS4.
and upgrading it means upgrading the rest of the jest pkgs.

**Comment by rebeccaalpert** [comment]
Closing this per Kim. I will work on fixing tests and resolving open comments in this PR in the new PR.
